foundry/foundry.py

```diff
@@ -365,12 +392,201 @@ def publish(self, foundry_metadata, data_source, title, authors, update=False,
         self.connect_client.add_organization(self.config.organization)
         self.connect_client.set_project_block(
             self.config.metadata_key, foundry_metadata)
-        self.connect_client.add_data_source(data_source)
+
+        # upload via HTTPS if specified
+        if https_data_path:
+            endpoint_id = "82f1b5c6-6e9b-11e5-ba47-22000b92c6ec"  # NCSA endpoint
+            # define upload destination
+            dest_path = self._create_dest_folder(endpoint_id)
+            # create new ACL rule (ie permission) for user to read/write to endpoint and path
+            rule_id = self._create_access_rule(endpoint_id, dest_path)
+            # upload data to endpoint
+            globus_data_source = self._https_upload(local_data_path=https_data_path, dest_path=dest_path,
+                                                    endpoint_id=endpoint_id)
+
+        # set Globus data source URL with MDF
+        self.connect_client.add_data_source(globus_data_source)
+        # set dataset name using the title if a an abbreviated short_name isn't specified
         self.connect_client.set_source_name(kwargs.get("short_name", title))
 
-        res = self.connect_client.submit_dataset(update=update)
+        # do not submit to MDF if this is just a test
+        if not test:
+            # Globus Transfer the data from the data source to the MDF endpoint
+            res = self.connect_client.submit_dataset(update=update)
+        else:
+            res = None
+
+        # if uploaded by HTTPS, delete ACL rule after dataset submission is complete
+        if https_data_path and rule_id:
+            self.transfer_client.delete_endpoint_acl_rule(endpoint_id, rule_id)
+
         return res
 
+    def _create_dest_folder(self, endpoint_id, parent_dir="/tmp"):
```

Comment:
Do we want to use type annotations through our library? I'm in favor as it will improve IDE experience

Reply:
Good point, I am still stuck in my old ways and forget about type annotations. I will add them ✨